### PR TITLE
Implement Lightweight OpenVPN API for Generating Client Certificates

### PR DESCRIPTION
**Pull request by goruha**
## what
* very simple/lightweight API for generating client certificates for OpenVPN

## why
* to be used with the Geodesic OpenVPN distribution that implements automatic OpenVPN client certificate generation using OAuth2. See: https://github.com/cloudposse/charts/incubator/openvpn/

## who
@cloudposse/engineering

### CODE REVIEW

File: .build-harness [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export SHELL = /bin/bash
```
Comment (osterman) [inline .build-harness]: Do not commit this file

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+script:
+- make go:deps
+- make go:test
+- make go:lint
+- make go:build-all
+
+deploy:
+  provider: releases
+  api_key: $GITHUB_OAUTH_TOKEN
```
Comment (osterman) [inline .travis.yml]: Use `GITHUB_API_KEY` for correctness.

File: src/api/openvpn.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+func GetClientConfig(name string) (result string, err error) {
+	err = nil
+	result = ""
+	err = recreateAllClientConfigs()
+	if err != nil {
+		return
+	}
+
+	file := "/etc/openvpn/clients/" + name + "/" + name + "-combined.ovpn"
```
Comment (osterman) [inline src/api/openvpn.go:41]: Lets make sure that `name` is alphanumeric.

File: src/api/openvpn.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			return true
+		}
+	}
+
+	return false
+}
+
+func cleanCertsFor(name string) (err error) {
+	err = nil
+
```
Comment (osterman) [inline src/api/openvpn.go]: Lets make sure that `name` is alphanumeric.

File: src/api/openvpn.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	if err = os.Remove(file); err != nil {
+		return
+	}
+
+	return
+
+}
+
+func generateCertsFor(user string) error {
+	cmd := exec.Command("easyrsa", "build-client-full", user, "nopass")
```
Comment (osterman) [inline src/api/openvpn.go]: Lets make sure that `user` is alphanumeric.

File: src/cmd/root.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+var cfgFile string
+
+var flags = []flag{
+	{"l", "string", "listen", ":8085", "Listen              ( environment variable LISTEN could be used instead )"},
+}
+
+// RootCmd represents the base command when called without any subcommands
+var RootCmd = &cobra.Command{
+	Use:   "openvpn-api",
+	Short: "Use Openvpn API to create users and fetch predefined openvpn configs",
```
Comment (osterman) [inline src/cmd/root.go]: OpenVPN

File: src/cmd/root.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	Long: `
+Use Openvpn API to create users and fetch predefined openvpn configs.
```
Comment (osterman) [inline src/cmd/root.go]: OpenVPN (use proper case)

File: src/server/server.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/cloudposse/openvpn-api/src/api"
+)
+
+// Run - start http server
+func Run(cfg config.Config) {
+
+	router := gin.Default()
+
+	router.GET("/user/:name", func(c *gin.Context) {
+		name := c.Param("name")
```
Comment (osterman) [inline src/server/server.go]: Lets make sure that `name` is alphanumeric.

**Review by osterman — CHANGES_REQUESTED** [review]
Remove `release/openvpn-api` from git

File: .gitignore [hunk 9]
```diff
@@ -22,3 +22,7 @@ _testmain.go
 *.exe
 *.test
 *.prof
+
+vendor
+.glide
+
```
Comment (osterman) [inline .gitignore]: add `.build-harness` here

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  file:
+    - release/$APP_darwin_386
```
Comment (osterman) [inline .travis.yml]: What do you think about using wildcards here? https://github.com/cloudposse/sudosh/blob/master/.travis.yml#L27

File: src/api/openvpn.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+		return
+	}
+
+	if err = deleteFileSafety("/etc/openvpn/pki/private/" + name + ".key"); err != nil {
+		return
+	}
+
+	return
+}
+
```
Comment (osterman) [inline src/api/openvpn.go:95]: Please annotate all these functions so it's clear what they do

File: .build-harness (hunk 1 above)
Comment (goruha) [reply .build-harness]: Done

File: .travis.yml (hunk 2 above)
Comment (goruha) [reply .travis.yml]: Done

File: .gitignore (hunk 9 above)
Comment (goruha) [reply .gitignore]: Done

File: .travis.yml (hunk 10 above)
Comment (goruha) [reply .travis.yml]: Nice.

Thanks.

File: src/cmd/root.go (hunk 7 above)
Comment (goruha) [reply src/cmd/root.go]: Done

File: src/cmd/root.go (hunk 6 above)
Comment (goruha) [reply src/cmd/root.go]: Done

File: src/server/server.go (hunk 8 above)
Comment (goruha) [reply src/server/server.go]: Done

File: src/api/openvpn.go (hunk 5 above)
Comment (goruha) [reply src/api/openvpn.go]: Done

File: src/api/openvpn.go (hunk 4 above)
Comment (goruha) [reply src/api/openvpn.go]: Done

File: src/api/openvpn.go (hunk 3 above)
Comment (goruha) [reply src/api/openvpn.go:41]: Done

File: src/api/openvpn.go (hunk 11 above)
Comment (goruha) [reply src/api/openvpn.go:95]: Done

File: src/api/openvpn.go [hunk 12]
```diff
@@ -87,6 +93,7 @@ func cleanCertsFor(name string) (err error) {
 	return
 }
 
+// Delete file if exists
 func deleteFileSafety(file string) (err error) {
```
Comment (osterman) [inline src/api/openvpn.go]: do you mean `deleteFileSafely` instead?

File: src/server/server.go [hunk 13]
```diff
@@ -5,15 +5,17 @@ import (
...
+	"github.com/dogenzaka/gin-tools/validation"
+	"github.com/dogenzaka/gin-tools/validation/validator"
 )
 
 // Run - start http server
 func Run(cfg config.Config) {
 
 	router := gin.Default()
-
+	router.Use(validation.ValidatePathParam("name", validator.RegExp{"^[a-zA-Z0-9]*$"}))
```
Comment (osterman) [inline src/server/server.go]: Name should contain at least 1 character. Change `*` to `+`.

**Review by osterman — CHANGES_REQUESTED** [review]
Please fix regex

File: src/api/openvpn.go (hunk 12 above)
Comment (goruha) [reply src/api/openvpn.go]: Right. 

typo.

Comment (goruha) [reply src/api/openvpn.go]: Done

File: src/server/server.go (hunk 13 above)
Comment (goruha) [reply src/server/server.go]: Done

**Review by osterman — APPROVED** [review]

**Pull request merged**